ops/pebble.py

```diff
@@ -752,5 +872,214 @@ def get_services(self, names: List[str] = None) -> List[ServiceInfo]:
         query = None
         if names is not None:
             query = {'names': ','.join(names)}
-        result = self._request('GET', '/v1/services', query)
-        return [ServiceInfo.from_dict(info) for info in result['result']]
+        resp = self._request('GET', '/v1/services', query)
+        return [ServiceInfo.from_dict(info) for info in resp['result']]
+
+    def read_content(self, path: str, encoding: str = 'utf-8') -> typing.Union[bytes, str]:
+        """Read a file's content from the remote system to a string.
+
+        Args:
+            path: Path of the file to read from the remote system.
+            encoding: Encoding to use for decoding file's bytes to a text string
+                (or None to return raw bytes).
+
+        Returns:
+            File's content, decoded and returned as str if encoding is not None,
+            or returned as raw bytes if encoding is None.
+        """
+        query = {
+            'action': 'read',
+            'path': path,
+        }
+        headers = {'Accept': 'multipart/form-data'}
+        response = self._request_raw('GET', '/v1/files', query, headers)
+
+        options = self._ensure_content_type(response.headers, 'multipart/form-data')
+        boundary = options.get('boundary', '')
+        if not boundary:
+            raise ProtocolError('invalid boundary {!r}'.format(boundary))
+
+        # We have to manually write the Content-Type with boundary, because
+        # email.parser expects the entire multipart message with headers.
+        parser = email.parser.BytesFeedParser()
+        parser.feed(b'Content-Type: multipart/form-data; boundary=' +
+                    boundary.encode('utf-8') + b'\r\n\r\n')
+
+        # Then read the rest of the response and feed it to the parser.
+        while True:
+            chunk = response.read(8192)
+            if not chunk:
+                break
+            parser.feed(chunk)
+        message = parser.close()
```

Comment:
Is there no support for streaming in the parser, or is it saving files in a temporary location similar to what Go does? We can't require the entire transference to be in memory. It's okay if we need to release v1 like that, but if that's what it takes we need to fix it shortly after.

Reply:
He did have an underlying API that was reading into a BinaryIO object, but it lead to a clumsy interaction if you just want to see the content of a config file.
I think it is one that we could introduce once we have a nice pattern (should it be read_file(path, IO) or read_file(path) => IO)

I think we started with read_content because we know that we want to have a simple way to get small content, and we didn't want to commit to an API for streaming content until we were comfortable with how it hung together.

Reply:
How about `push(path, file_like)` and `pull(path) => file_like`?  `pull(path).read()` seems short and idiomatic.

Reply:
@niemeyer Yeah, that looks like a nice API at first glance, but 1) it's tricky to implement, as the file_like returned from `pull` would have to parse multipart in a streaming fashion -- there's no direct support for what we need in the stdlib -- and then do something on the last `file_like.read()` call to process the JSON metadata in the last multipart part, 2) because of that, it means that `pull` itself won't raise errors with the overall call, but the last `read()` may raise errors, which seems a bit weird to me.

On balance I do still think that's probably a reasonable API, but it'd take some time to implement correctly, so we would like to do it separately later. I can play with it some more if you feel strongly that we *don't* want `read_content` or `write_content`.

Reply:
After discussing with the team, we concluded we'll want string-returning and string-accepting functions anyway, so I'm going ahead with `read_content` and `write_content` for v1. We'll plan to add file-like object variants later -- likely named `push` and `pull`, once we settle on the right API and error handling.